File: codes/python/ch08/bvp_nonlinear.py
#!/usr/bin/env python3
"""
bvp_nonlinear.py

Solves the Bratu equation (nonlinear BVP) using Newton iteration:

    u_xx + λ e^u = 0,  x ∈ (-1, 1),  u(±1) = 0

with λ = 0.5 (subcritical case, unique solution exists).

IMPORTANT: The critical value for [-1, 1] domain is λ_c ≈ 0.878.
For λ > λ_c, no solution exists! We use λ = 0.5 for a stable solution.

This demonstrates that nonlinear BVPs are easily handled with spectral
methods using Newton iteration.

This script generates Figure 7.5 and Table 7.1 for Chapter 7.

Author: Dr. Denys Dutykh
        Mathematics Department
        Khalifa University of Science and Technology
        Abu Dhabi, UAE

Part of the book "Computational Etudes: A Spectral Approach"
https://github.com/dutykh/computational-etudes
"""

import numpy as np
import matplotlib.pyplot as plt
from matplotlib import rcParams
from pathlib import Path

# Import Chebyshev matrix function from Chapter 6
import sys
sys.path.insert(0, str(Path(__file__).parent.parent / 'ch07'))
from cheb_matrix import cheb_matrix, cheb_second_derivative_matrix

# -----------------------------------------------------------------------------
# Publication-quality matplotlib configuration
# -----------------------------------------------------------------------------
rcParams['font.family'] = 'serif'
rcParams['font.serif'] = ['Computer Modern Roman', 'CMU Serif', 'DejaVu Serif']
rcParams['mathtext.fontset'] = 'cm'
rcParams['axes.labelsize'] = 11
rcParams['axes.titlesize'] = 12
rcParams['xtick.labelsize'] = 10
rcParams['ytick.labelsize'] = 10
rcParams['legend.fontsize'] = 10
rcParams['figure.dpi'] = 150
rcParams['savefig.dpi'] = 300
rcParams['text.usetex'] = False
rcParams['axes.linewidth'] = 0.8
rcParams['xtick.major.width'] = 0.8
rcParams['ytick.major.width'] = 0.8

# -----------------------------------------------------------------------------
# Book color scheme
# -----------------------------------------------------------------------------
NAVY = '#142D6E'
SKY = '#7896D2'
CORAL = '#E74C3C'
TEAL = '#16A085'
PURPLE = '#8E44AD'
ORANGE = '#E67E22'

# Output paths
SCRIPT_DIR = Path(__file__).parent
OUTPUT_DIR = SCRIPT_DIR / '..' / '..' / '..' / 'textbook' / 'figures' / 'ch08' / 'python'


def solve_bratu_newton(N, lam=0.5, tol=1e-12, max_iter=30):
    """
    Solve the Bratu equation u_xx + λ e^u = 0 with u(±1) = 0 using Newton.

    Newton iteration for F(u) = D²u + λ e^u = 0:
        J δu = -F(u)
        J = D² + λ diag(e^u)
        u_new = u + α δu  (with line search for robustness)

    Parameters
    ----------
    N : int
        Number of intervals
    lam : float
        Parameter λ (default 1.0)
    tol : float
        Convergence tolerance
    max_iter : int
        Maximum iterations

    Returns
    -------
    x : ndarray
        Chebyshev grid points
    u : ndarray
        Solution
    residuals : list
        Convergence history (residual norms)
    n_iter : int
        Number of iterations used
    """
    D2, D, x = cheb_second_derivative_matrix(N)

    # Extract interior system
    D2_int = D2[1:N, 1:N]

    # Initial guess: u = 0
    u = np.zeros(N + 1)

    residuals = []

    for it in range(max_iter):
        # Residual: F(u) = D²u + λ e^u (interior points)
        exp_u = np.exp(u[1:N])
        F = D2_int @ u[1:N] + lam * exp_u

        res_norm = np.max(np.abs(F))
        residuals.append(res_norm)

        if res_norm < tol:
            return x, u, residuals, it + 1

        # Jacobian: J = D² + λ diag(e^u)
        J = D2_int + lam * np.diag(exp_u)

        # Newton step
        delta_u = np.linalg.solve(J, -F)

        # Backtracking line search to ensure residual decreases
        alpha = 1.0
        u_trial = u.copy()
        for _ in range(10):  # Max 10 backtracking steps
            u_trial[1:N] = u[1:N] + alpha * delta_u
            exp_u_trial = np.exp(u_trial[1:N])
            F_trial = D2_int @ u_trial[1:N] + lam * exp_u_trial
            res_trial = np.max(np.abs(F_trial))

            if res_trial < res_norm:
                break
            alpha *= 0.5

        # Update interior points with damped step
        u[1:N] = u[1:N] + alpha * delta_u

    return x, u, residuals, max_iter


def solve_bratu_fixedpoint(N, lam=0.5, tol=1e-10, max_iter=100):
    """
    Solve Bratu equation using fixed-point iteration for comparison.

    Fixed-point iteration: D² u_new = -λ exp(u_old)
    """
    D2, D, x = cheb_second_derivative_matrix(N)
    D2_int = D2[1:N, 1:N]

    # Initial guess
    u = np.zeros(N + 1)

    residuals = []

    for it in range(max_iter):
        # RHS
        rhs = -lam * np.exp(u[1:N])

        # Solve
        u_new_int = np.linalg.solve(D2_int, rhs)

        # Compute change
        change = np.max(np.abs(u_new_int - u[1:N]))
        residuals.append(change)

        u[1:N] = u_new_int

        if change < tol:
            return x, u, residuals, it + 1

    return x, u, residuals, max_iter


def main():
    """Create nonlinear BVP figure."""

    OUTPUT_DIR.mkdir(parents=True, exist_ok=True)

    # Create figure
    fig, axes = plt.subplots(1, 2, figsize=(11, 4.5))

    # Panel 1: Solution for various N
    ax1 = axes[0]

    N_values = [8, 16, 32]
    colors = [CORAL, TEAL, PURPLE]

    for N, color in zip(N_values, colors):
        x, u, residuals, n_iter = solve_bratu_newton(N)
        ax1.plot(x, u, 'o-', color=color, linewidth=1.2, markersize=4,
                 markeredgecolor='white', markeredgewidth=0.3,
                 label=f'$N = {N}$ ({n_iter} iter)')

    ax1.set_xlabel(r'$x$')
    ax1.set_ylabel(r'$u(x)$')
    ax1.set_title(r'Bratu Equation: $u_{xx} + e^u = 0$', fontsize=11)
    ax1.legend(loc='lower center', fontsize=9, ncol=3)
    ax1.set_xlim(-1.05, 1.05)
    ax1.spines['top'].set_visible(False)
    ax1.spines['right'].set_visible(False)
    ax1.yaxis.grid(True, linestyle='-', alpha=0.2, color='#888888')
    ax1.xaxis.grid(True, linestyle='-', alpha=0.2, color='#888888')

    # Panel 2: Convergence history
    ax2 = axes[1]

    # Compare fixed-point vs Newton for N=16
    N = 16
    _, _, res_fp, n_fp = solve_bratu_fixedpoint(N, tol=1e-14, max_iter=50)
    _, _, res_newton, n_newton = solve_bratu_newton(N, tol=1e-14, max_iter=20)

    ax2.semilogy(range(1, len(res_fp) + 1), res_fp, 'o-', color=CORAL,
                 linewidth=1.5, markersize=5, label='Fixed-point')
    ax2.semilogy(range(1, len(res_newton) + 1), res_newton, 's-', color=TEAL,
                 linewidth=1.5, markersize=5, label='Newton')

    ax2.set_xlabel('Iteration')
    ax2.set_ylabel('Residual norm')
    ax2.set_title(f'Convergence History ($N = {N}$)', fontsize=11)
    ax2.legend(loc='upper right', fontsize=9)
    ax2.grid(True, which='both', alpha=0.3)
    ax2.spines['top'].set_visible(False)
    ax2.spines['right'].set_visible(False)
    ax2.set_ylim(1e-16, 10)

    # Add tolerance line
    ax2.axhline(1e-12, color='gray', linestyle='--', linewidth=1, alpha=0.7)
    ax2.text(2, 3e-12, 'tol', fontsize=9, color='gray')

    # Main title
    fig.suptitle(r'Nonlinear BVP: Bratu Equation with $\lambda = 0.5$',
                 fontsize=13, y=0.98)

    plt.tight_layout(rect=[0, 0, 1, 0.94])

    # Save figure
    output_file = OUTPUT_DIR / 'bratu.pdf'
    fig.savefig(output_file, bbox_inches='tight', pad_inches=0.05)
    fig.savefig(output_file.with_suffix('.png'), bbox_inches='tight',
                pad_inches=0.05, dpi=300)

    print(f'Figure saved to: {output_file.resolve()}')
    plt.close(fig)

    # Generate convergence table
    print("\n" + "=" * 60)
    print("Table 7.1: Bratu Equation Convergence (Newton)")
    print("=" * 60)
    print(f"{'N':>6} {'Iterations':>12} {'u(0)':>16}")
    print("-" * 60)

    for N in [4, 8, 16, 32, 64]:
        x, u, _, n_iter = solve_bratu_newton(N)
        # Find index closest to x=0
        idx_0 = np.argmin(np.abs(x))
        print(f"{N:>6d} {n_iter:>12d} {u[idx_0]:>16.10f}")

    print("=" * 60)
    print("\nNote: Newton iteration converges quadratically (in 5-8 iterations)")
    print("while fixed-point converges linearly (in ~18 iterations).")


if __name__ == '__main__':
    main()

File: codes/julia/ch08/bvp_nonlinear.jl
# bvp_nonlinear.jl
#
# Solves the Bratu equation (nonlinear BVP) using Newton iteration:
#
#     u_xx + lambda * exp(u) = 0,  x in (-1, 1),  u(+/-1) = 0
#
# with lambda = 0.5 (subcritical case, unique solution exists).
#
# IMPORTANT: The critical value for [-1, 1] domain is lambda_c ~ 0.878.
# For lambda > lambda_c, no solution exists! We use lambda = 0.5 for a
# stable solution.
#
# This demonstrates that nonlinear BVPs are easily handled with spectral
# methods using Newton iteration.
#
# Author: Dr. Denys Dutykh
#         Mathematics Department
#         Khalifa University of Science and Technology
#         Abu Dhabi, UAE
#
# Part of the book "Computational Etudes: A Spectral Approach"
# https://github.com/dutykh/computational-etudes

using CairoMakie
using Colors
using LinearAlgebra
using Printf

# Import Chebyshev matrix function from Chapter 7
include(joinpath(@__DIR__, "..", "ch07", "cheb_matrix.jl"))

# -----------------------------------------------------------------------------
# Publication-quality CairoMakie configuration
# -----------------------------------------------------------------------------
set_theme!(Theme(
    fontsize = 11,
    fonts = (regular = "CMU Serif", bold = "CMU Serif Bold", italic = "CMU Serif Italic"),
    Axis = (xlabelsize = 11, ylabelsize = 11, titlesize = 12,
            xticklabelsize = 10, yticklabelsize = 10,
            spinewidth = 0.8, xtickwidth = 0.8, ytickwidth = 0.8),
    Legend = (labelsize = 10,),
))

# -----------------------------------------------------------------------------
# Book color scheme
# -----------------------------------------------------------------------------
const NAVY   = colorant"#142D6E"
const SKY    = colorant"#7896D2"
const CORAL  = colorant"#E74C3C"
const TEAL   = colorant"#16A085"
const PURPLE = colorant"#8E44AD"
const ORANGE = colorant"#E67E22"

# Output paths
const SCRIPT_DIR = @__DIR__
const OUTPUT_DIR = joinpath(SCRIPT_DIR, "..", "..", "..", "textbook", "figures", "ch08", "julia")

# -----------------------------------------------------------------------------
# Solvers
# -----------------------------------------------------------------------------
"""
    solve_bratu_newton(N; lam=0.5, tol=1e-12, max_iter=30)

Solve the Bratu equation u_xx + lambda*exp(u) = 0 with u(+/-1) = 0 using Newton.

Newton iteration for F(u) = D^2*u + lambda*exp(u) = 0:
    J*delta_u = -F(u)
    J = D^2 + lambda*diag(exp(u))
    u_new = u + alpha*delta_u  (with line search for robustness)
"""
function solve_bratu_newton(N; lam=0.5, tol=1e-12, max_iter=30)
    D2, _, x = cheb_second_derivative_matrix(N)

    # Extract interior system
    D2_int = D2[2:N, 2:N]

    # Initial guess: u = 0
    u = zeros(N + 1)

    residuals = Float64[]

    for it in 1:max_iter
        # Residual: F(u) = D^2*u + lambda*exp(u) (interior points)
        exp_u = exp.(u[2:N])
        F = D2_int * u[2:N] + lam * exp_u

        res_norm = maximum(abs.(F))
        push!(residuals, res_norm)

        if res_norm < tol
            return x, u, residuals, it
        end

        # Jacobian: J = D^2 + lambda*diag(exp(u))
        J = D2_int + lam * Diagonal(exp_u)

        # Newton step
        delta_u = J \ (-F)

        # Backtracking line search to ensure residual decreases
        alpha = 1.0
        u_trial = copy(u)
        for _ in 1:10    # Max 10 backtracking steps
            u_trial[2:N] = u[2:N] + alpha * delta_u
            exp_u_trial = exp.(u_trial[2:N])
            F_trial = D2_int * u_trial[2:N] + lam * exp_u_trial
            res_trial = maximum(abs.(F_trial))

            if res_trial < res_norm
                break
            end
            alpha *= 0.5
        end

        # Update interior points with damped step
        u[2:N] = u[2:N] + alpha * delta_u
    end

    return x, u, residuals, max_iter
end

"""
    solve_bratu_fixedpoint(N; lam=0.5, tol=1e-10, max_iter=100)

Solve Bratu equation using fixed-point iteration for comparison.
Fixed-point iteration: D^2 u_new = -lambda*exp(u_old)
"""
function solve_bratu_fixedpoint(N; lam=0.5, tol=1e-10, max_iter=100)
    D2, _, x = cheb_second_derivative_matrix(N)
    D2_int = D2[2:N, 2:N]

    # Initial guess
    u = zeros(N + 1)

    residuals = Float64[]

    for it in 1:max_iter
        # RHS
        rhs = -lam * exp.(u[2:N])

        # Solve
        u_new_int = D2_int \ rhs

        # Compute change
        change = maximum(abs.(u_new_int .- u[2:N]))
        push!(residuals, change)

        u[2:N] = u_new_int

        if change < tol
            return x, u, residuals, it
        end
    end

    return x, u, residuals, max_iter
end

function main()
    mkpath(OUTPUT_DIR)

    # Create figure
    fig = Figure(size=(950, 400))

    # Panel 1: Solution for various N
    ax1 = Axis(fig[1, 1],
        xlabel = L"x",
        ylabel = L"u(x)",
        title = L"Bratu Equation: $u_{xx} + \lambda e^u = 0$",
        titlesize = 11)

    N_values = [8, 16, 32]
    colors = [CORAL, TEAL, PURPLE]

    for (N, color) in zip(N_values, colors)
        x, u, residuals, n_iter = solve_bratu_newton(N)
        scatterlines!(ax1, x, u, color=color, linewidth=1.2, markersize=5,
                      strokecolor=:white, strokewidth=0.3,
                      label=@sprintf("N = %d (%d iter)", N, n_iter))
    end

    xlims!(ax1, -1.05, 1.05)
    axislegend(ax1, position=:cb, labelsize=9, nbanks=3)
    hidespines!(ax1, :t, :r)

    # Panel 2: Convergence history
    ax2 = Axis(fig[1, 2],
        xlabel = "Iteration",
        ylabel = "Residual norm",
        title = "Convergence History (N = 16)",
        titlesize = 11,
        yscale = log10)

    # Compare fixed-point vs Newton for N=16
    N = 16
    _, _, res_fp, n_fp = solve_bratu_fixedpoint(N, tol=1e-14, max_iter=50)
    _, _, res_newton, n_newton = solve_bratu_newton(N, tol=1e-14, max_iter=20)

    scatterlines!(ax2, 1:length(res_fp), res_fp, color=CORAL,
                  linewidth=1.5, markersize=7, label="Fixed-point")
    scatterlines!(ax2, 1:length(res_newton), res_newton, color=TEAL,
                  linewidth=1.5, markersize=7, marker=:rect, label="Newton")

    ylims!(ax2, 1e-16, 10)
    axislegend(ax2, position=:rt, labelsize=9)
    hidespines!(ax2, :t, :r)

    # Add tolerance line
    hlines!(ax2, [1e-12], color=:gray, linestyle=:dash, linewidth=1)
    text!(ax2, 2, 3e-12, text="tol", fontsize=9, color=:gray)

    # Main title
    Label(fig[0, :],
        L"Nonlinear BVP: Bratu Equation with $\lambda = 0.5$",
        fontsize=13)

    # Save figure
    save(joinpath(OUTPUT_DIR, "bratu.pdf"), fig)
    save(joinpath(OUTPUT_DIR, "bratu.png"), fig, px_per_unit=4)

    println("Figure saved to: ", joinpath(OUTPUT_DIR, "bratu.pdf"))

    # Generate convergence table
    println("\n" * "=" ^ 60)
    println("Table 7.1: Bratu Equation Convergence (Newton)")
    println("=" ^ 60)
    @printf("%6s %12s %16s\n", "N", "Iterations", "u(0)")
    println("-" ^ 60)

    for N in [4, 8, 16, 32, 64]
        x, u, _, n_iter = solve_bratu_newton(N)
        # Find index closest to x=0
        idx_0 = argmin(abs.(x))
        @printf("%6d %12d %16.10f\n", N, n_iter, u[idx_0])
    end

    println("=" ^ 60)
    println("\nNote: Newton iteration converges quadratically (in 5-8 iterations)")
    println("while fixed-point converges linearly (in ~18 iterations).")
end

main()
